inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/scheduling/tasks/PredictionTask.java

```diff
@@ -62,66 +77,95 @@ public void run()
+                annoService.upgradeCas(jCas.getCas(), document, user.getUsername());
+            } catch (IOException e) {
+                log.error("Cannot read annotation CAS.", e);
                 continue;
-            }
-            
-            List<Recommender> recommenders = recommendationService.getActiveRecommenders(user,
-                    layer);
-
-            if (recommenders.isEmpty()) {
-                log.debug("[{}][{}]: No active recommenders, skipping prediction.",
-                        user.getUsername(), layer.getUiName());
+            } catch (UIMAException e) {
+                log.error("Cannot upgrade annotation CAS.", e);
                 continue;
             }
-            
-            for (Recommender recommender : recommenders) {
-                long startTime = System.currentTimeMillis();
-                
-                ClassificationTool<?> ct = recommendationService.getTool(recommender,
-                        recommendationService.getMaxSuggestions(user));
-                Classifier<?> classifier = ct.getClassifier();
-
-                classifier.setUser(getUser());
-                classifier.setProject(getProject());
-                classifier.setModel(recommendationService.getTrainedModel(user, recommender));
-    
-                List<AnnotationObject> predictions = new ArrayList<>();
-                
-                log.info("[{}][{}]: Predicting labels...", user.getUsername(),
-                        recommender.getName());
-                
-                List<AnnotationDocument> docs = documentService
-                        .listAnnotationDocuments(layer.getProject(), user);
-
-                docs.forEach(doc -> {
+
+            for (AnnotationLayer layer : annoService.listAnnotationLayer(project)) {
+                if (!layer.isEnabled()) {
+                    continue;
+                }
+
+                List<Recommender> recommenders = recommendationService
+                    .getActiveRecommenders(user, layer);
+
+                for (Recommender recommender : recommenders) {
+                    RecommenderContext ctx = recommendationService.getContext(user, recommender);
+                    RecommendationEngineFactory factory = recommendationService
+                            .getRecommenderFactory(recommender);
+                    RecommendationEngine recommendationEngine = factory.build(recommender);
+
                     try {
-                        JCas jcas = documentService.readAnnotationCas(doc);
-                        predictions.addAll(classifier.predict(jcas, layer));
-                    } catch (IOException e) {
-                        log.error("Cannot read annotation CAS.", e);
+                        recommendationEngine.predict(ctx, jCas.getCas());
+                    } catch (RecommendationException e) {
+                        log.error("Error while predicting", e);
```

Comment:
Error message should include information on which recommender/document/project the problem happened and whether that causes the entire prediction to fail or only the document to be skipped (or the remainder of the document to be skipped?).